biahub/track.py

```diff
@@ -0,0 +1,694 @@
+
+    Parameters:
+    - data_dict (dict): Dictionary containing imaging data with channel names as keys and
+                         corresponding NumPy arrays as values.
+    - preprocessing_functions (Dict[str, ProcessingFunctions]): Dictionary mapping preprocessing
+                                                                steps to their corresponding functions.
+    - tracking_functions (Dict[str, ProcessingFunctions]): Dictionary containing functions for
+                                                            generating foreground and contour masks.
+
+    Returns:
+    - Tuple[ArrayLike, ArrayLike]:
+        - The foreground mask used for tracking.
+        - The contour gradient map used for tracking.
+
+    Notes:
+    - If the "Phase3D" key is present in `data_dict`, the function identifies and fills empty frames
+      before processing.
+    - The preprocessing functions are dynamically resolved and applied to the corresponding
+      input arrays.
+    """
+
+    click.echo("Checking for empty frames in the label-free image...")
+
+    if empty_frames_idx:
+        for key, value in data_dict.items():
+            click.echo(f"Filling empty frames in {key}...")
+            data_dict[key] = fill_empty_frames(value, empty_frames_idx)
+
+    # Apply preprocessing functions
+    for key, func_details in preprocessing_functions.items():
+        click.echo(f"Preprocessing {key} using {func_details.function}...")
+        input_arrays_name = func_details.input_arrays[0]
+        function = resolve_function(func_details.function)  # Uses function mapping
+        input_array = data_dict[input_arrays_name]
+
+        kwargs = func_details.kwargs if func_details.kwargs else {}
+        data_dict[input_arrays_name] = array_apply(input_array, func=function, **kwargs)
+    # Generate foreground mask
+    click.echo("Generating foreground mask...")
+    foreground_func_details = tracking_functions["foreground"]
+    foreground_function = resolve_function(
+        foreground_func_details.function
+    )  # Uses function mapping
+
+    fg_input_arrays = [data_dict[name] for name in foreground_func_details.input_arrays]
+    fg_kwargs = foreground_func_details.kwargs if foreground_func_details.kwargs else {}
+
+    foreground_mask = array_apply(*fg_input_arrays, func=foreground_function, **fg_kwargs)
+
+    # Generate contour gradient map
+    click.echo("Generating contour gradient map...")
+    contour_func_details = tracking_functions["contour"]
+    contour_function = resolve_function(contour_func_details.function)  # Uses function mapping
+
+    contour_input_arrays = [data_dict[name] for name in contour_func_details.input_arrays]
+    contour_kwargs = contour_func_details.kwargs if contour_func_details.kwargs else {}
+
+    contour_gradient_map = array_apply(
+        *contour_input_arrays, func=contour_function, **contour_kwargs
+    )
+
+    return foreground_mask, contour_gradient_map
+
+
+def ultrack(
```

Comment:
This may cause confusion if `ultrack`-the-package is also imported in the call site's scope.

Reply:
LSP and other static analysis tools can also get confused.